### PR TITLE
connection-pool-refactoring-all-changes

### PR DESCRIPTION
**Pull request by elopezcastro**
Summary of changes:

- Fixed issue with LocalH2 data source identifier being too long and clogging logs
- Fixed issue with creating unnecessary hikary pools for every connection request for LocalH2 and Embedded data sources
- Consolidated pool state : we just have a single state of connection pools ConnectionStateManager
- Pool management responsibility :  ConnectionStateManager deals with connection pool status and requests , pool locks are safer. This functionality has been moved from DataSourceSpecification
- Connection pool Housekeeping: unused connection pools areproperly closed and disposed by housekeeper task
- Added the ability to configure max pool and min pool sizes per DataSource type

### CODE REVIEW

File: legend-engine-executionPlan-execution-store-relational-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/relational/connection/ds/state/ConnectionStateManager.java [hunk 1]
```diff
@@ -244,47 +317,114 @@ public ConnectionStateEvictionTask(long durationInSeconds)
...
+        Function0<DataSourceWithStatistics> dsSupplier = () -> new DataSourceWithStatistics(poolName,identityState,dataSourceSpecification);
+        DataSource dataSource = this.connectionPools.getIfAbsentPut(poolName, dsSupplier).getDataSource();
+        if (dataSource == null)
+        {
+            synchronized (poolLockManager.getLock(poolName))
+            {
+                dataSource = this.connectionPools.getIfAbsentPut(poolName, dsSupplier).getDataSource();
+                if (dataSource == null)
+                {
+                    LOGGER.info("Pool not found for [{}] for datasource [{}], creating one", principal, connectionKey.shortId());
```
Comment (epsstan) [inline legend-engine-executionPlan-execution-store-relational-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/relational/connection/ds/state/ConnectionStateManager.java:347]: :confused: I don't follow this synchronization code. 

Lines 339 and 344 use the map's lock (which is tied to the poolName). In addition line 342 uses an explicit lock also tied to the poolName.

In the previous implementation (when this code was in DataSourceSpecification.java), the locking idiom was 
```
object = map.get(key)
if (object == null)
{
   synchronized(makeLock(key))
  {
     object = map.get(key)
     if (object == null) { map.put(key, makeObject()) } 
  }
}
```
If getIfAbsentPut is atomic, do we need the null checks and the explicit lock ?

File: pom.xml [hunk 2]
```diff
@@ -125,7 +125,7 @@
         <maven.compiler.release>8</maven.compiler.release>
 
         <!-- Databases -->
-        <hikaricp.version>3.4.5</hikaricp.version>
+        <hikaricp.version>4.0.3</hikaricp.version>
```
Comment (epsstan) [inline pom.xml:128]: :worried: 

Do we need to pickup this version upgrade now. 
There are already a lot of changes in this PR. Let's skip this if we absolutely do not need it for this PR.

File: legend-engine-executionPlan-execution-store-relational-connection-tests/src/test/java/org/finos/legend/engine/plan/execution/stores/relational/connection/test/TestConnectionState.java [hunk 3]
```diff
@@ -95,15 +95,15 @@ public void connectionStateUpdate() throws Exception
 
         //verify connection state for user1 exists
         String poolName = String.format("DBPool_Static_host:127.0.0.1_port:%d_db:db1_type:TestDB_testuser1", server.getPort());
-        ConnectionState connectionState1 = ConnectionStateManager.getInstance().getState(poolName);
+        DataSourceWithStatistics connectionState1 = ConnectionStateManager.getInstance().getDataSourceByPoolName(poolName);
```
[the remark this answers is not in the document]
Comment (elopezcastro) [reply legend-engine-executionPlan-execution-store-relational-connection-tests/src/test/java/org/finos/legend/engine/plan/execution/stores/relational/connection/test/TestConnectionState.java:98]: done

File: legend-engine-executionPlan-execution-store-relational-connection-tests/src/test/java/org/finos/legend/engine/plan/execution/stores/relational/connection/test/TestLocalH2ConnectionCreation.java [hunk 4]
```diff
@@ -125,39 +163,90 @@ public void multipleUsersAcquireConnectionsToDifferentDatabases() throws Excepti
...
+        Assert.assertNotNull(key2);
+
+        Assert.assertEquals(key1, key2);
 
-        // We do not have a connection pool for either user
-        assertEquals(0, ConnectionPoolTestUtils.countNumHikariPools(identity1.getName()));
-        assertEquals(0, ConnectionPoolTestUtils.countNumHikariPools(identity2.getName()));
+
+        Assert.assertNotNull(this.connectionStateManager.get(poolName1));
+        Assert.assertNotNull(this.connectionStateManager.get(poolName2));
```
[the remark this answers is not in the document]
Comment (elopezcastro) [reply legend-engine-executionPlan-execution-store-relational-connection-tests/src/test/java/org/finos/legend/engine/plan/execution/stores/relational/connection/test/TestLocalH2ConnectionCreation.java:225]: done

File: legend-engine-executionPlan-execution-store-relational-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/relational/connection/ds/DataSourceSpecification.java [hunk 5]
```diff
@@ -154,95 +127,60 @@ public Connection getConnectionUsingSubject(Subject subject)
...
             {
-                int requests = dataSourceWithStatistics.requestConnection();
+                DataSourceWithStatistics dataSourceWithStatistics = this.connectionStateManager.getDataSourceForIdentityIfAbsentBuild(identityState,this,dataSourcePoolBuilder);
                 // Logs and traces and stats -----
-                scope.span().setTag("Pool", dataSourceWithStatistics.getDataSource().toString());
-                String poolName = getPoolName(dataSourceWithStatistics.getDataSource());
-                LOGGER.info("Found pool for [{}] in datasource [{}] : pool Name [{}]", principal, dataSourceWithStatistics.getDataSource(), poolName);
+                String poolName = dataSourceWithStatistics.getPoolName();
+                scope.span().setTag("Pool", poolName);
+                int requests = dataSourceWithStatistics.requestConnection();
```
[the remark this answers is not in the document]
Comment (elopezcastro) [reply legend-engine-executionPlan-execution-store-relational-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/relational/connection/ds/DataSourceSpecification.java:160]: have moved to somewhere else

File: legend-engine-executionPlan-execution-store-relational-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/relational/connection/ds/DataSourceWithStatistics.java [hunk 6]
```diff
@@ -40,6 +65,68 @@ public int requestConnection()
...
+    }
+
+    public Identity getIdentity()
+    {
+        return identityState.getIdentity();
+    }
+
+    public Object getCredentialSupplier()
+    {
+        return identityState.getCredentialSupplier();
```
[the remark this answers is not in the document]
Comment (elopezcastro) [reply legend-engine-executionPlan-execution-store-relational-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/relational/connection/ds/DataSourceWithStatistics.java:103]: done

File: legend-engine-executionPlan-execution-store-relational-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/relational/connection/ds/state/ConnectionStateManager.java [hunk 7]
```diff
@@ -141,101 +167,148 @@ private static ConnectionStateManager getInstanceImpl(Clock clock)
...
     {
         int sizeBeforePurge = this.size();
-        LOGGER.info("Connection state purge : Starting purge with cache size={}", sizeBeforePurge);
+        LOGGER.info("DataSource state : Starting  with cache size={}", sizeBeforePurge);
         this.evictStateOlderThan(Duration.ofSeconds(durationInSeconds));
         int sizeAfterPurge = this.size();
-        LOGGER.info("Connection state purge : Evicted={}", sizeBeforePurge-sizeAfterPurge);
-        this.dump();
-        LOGGER.info("Connection state purge : Completed purge with cache size={}", sizeAfterPurge);
+        LOGGER.info("Datasource state : Evicted={}", sizeBeforePurge - sizeAfterPurge);
```
[the remark this answers is not in the document]
Comment (elopezcastro) [reply legend-engine-executionPlan-execution-store-relational-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/relational/connection/ds/state/ConnectionStateManager.java:275]: corrected

File: legend-engine-executionPlan-execution-store-relational-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/relational/connection/ds/state/ConnectionStateManager.java [hunk 8]
```diff
@@ -244,47 +317,114 @@ public ConnectionStateEvictionTask(long durationInSeconds)
...
         {
             ConnectionStateManager instance = ConnectionStateManager.getInstance();
-            instance.purge(durationInSeconds);
+            try
+            {
+                instance.purge(durationInSeconds);
+            }
+            catch (Exception e)
+            {
+                LOGGER.error("Datasource Connection State purge failed {}", e);
```
[the remark this answers is not in the document]
Comment (elopezcastro) [reply legend-engine-executionPlan-execution-store-relational-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/relational/connection/ds/state/ConnectionStateManager.java:326]: corrected

File: pom.xml (hunk 2 above)
Comment (elopezcastro) [reply pom.xml:128]: have done extensive testing

File: legend-engine-executionPlan-execution-store-relational-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/relational/connection/ds/state/ConnectionStateManager.java (hunk 1 above)
Comment (elopezcastro) [reply legend-engine-executionPlan-execution-store-relational-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/relational/connection/ds/state/ConnectionStateManager.java:347]: discussed offline will add more commentary to help developers understand

File: legend-engine-executionPlan-execution-store-relational-connection-tests/src/test/java/org/finos/legend/engine/plan/execution/stores/relational/connection/test/TestLocalH2ConnectionCreation.java [hunk 9]
```diff
@@ -86,33 +92,65 @@ public void userAcquiresConcurrentConnectionsToSameDb() throws Exception
...
+
+        // We have a connection pool for the user
+
+        assertNotNull(this.connectionStateManager.get(this.connectionStateManager.poolNameFor(identity1,connectionKey)));
+
+        H2TestUtils.closeProperly(db1Conn1, db2Conn1);
+    }
+
+    @Test
+    public void userAcquiresConcurrentConnectionsToDifferentDbs() throws Exception
```
[the remark this answers is not in the document]
Comment (elopezcastro) [reply legend-engine-executionPlan-execution-store-relational-connection-tests/src/test/java/org/finos/legend/engine/plan/execution/stores/relational/connection/test/TestLocalH2ConnectionCreation.java:133]: this test is differennt, this is testing we get correct results when we use same datasouce specifications

File: legend-engine-executionPlan-execution-store-relational-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/relational/connection/ds/state/ConnectionStateManager.java [hunk 10]
```diff
@@ -87,27 +107,28 @@ This is generally (??) safe. Since the racing threads are for the same pool and
...
+    private static final String SEPARATOR = "_";
 
     private static ConnectionStateManager INSTANCE;
 
     static
     {
         ThreadFactory threadFactory = r -> new Thread(r, "ConnectionStateManager Housekeeper");
         long evictionDurationInSeconds = resolveEvictionDuration();
-        ConnectionStateEvictionTask connectionStateEvictionTask = new ConnectionStateEvictionTask(evictionDurationInSeconds);
+        ConnectionStateHousekeepingTask connectionStateEvictionTask = new ConnectionStateHousekeepingTask(evictionDurationInSeconds);
```
[the remark this answers is not in the document]
Comment (elopezcastro) [reply legend-engine-executionPlan-execution-store-relational-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/relational/connection/ds/state/ConnectionStateManager.java:125]: done

File: legend-engine-executionPlan-execution-store-relational-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/relational/connection/ds/specifications/LocalH2DataSourceSpecification.java [hunk 11]
```diff
@@ -14,55 +14,40 @@
...
-            AlloyH2Server.startServer(port);
-        }
-        catch (SQLException e)
-        {
-            throw new RuntimeException(e);
-        }
-    }
 
-    public LocalH2DataSourceSpecification(LocalH2DataSourceSpecificationKey key, DatabaseManager databaseManager, AuthenticationStrategy authenticationStrategy, RelationalExecutorInfo relationalExecutorInfo)
+    private static final int MAX_POOL_SIZE = 100;
```
[the remark this answers is not in the document]
Comment (elopezcastro) [reply legend-engine-executionPlan-execution-store-relational-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/relational/connection/ds/specifications/LocalH2DataSourceSpecification.java:35]: keeping to 100 for now which is current behavour, have noticed we have tests failures when running mutiple paralell builds

File: legend-engine-executionPlan-execution-store-relational-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/relational/connection/ds/state/ConnectionStateManager.java [hunk 12]
```diff
@@ -141,101 +167,148 @@ private static ConnectionStateManager getInstanceImpl(Clock clock)
...
-        for (Map.Entry<String, ConnectionState> entry : entriesToPurge)
-        {
-            String poolName = entry.getKey();
-            ConnectionState state = entry.getValue();
-            // step 2 - remove atomically - i.e remove iff the state has not been updated since it was read in step 1
-            this.atomicallyRemove(poolName, state);
-        }
+        // step 1 - gather pools to be deleted without acquiring a global lock
+        Set<Pair<String, DataSourceStatistics>> entriesToPurge = this.findPoolsOlderThan(duration);
+        LOGGER.info(" Datasource state purgePoolStates : pool {} to be evicted", entriesToPurge.size());
```
[the remark this answers is not in the document]
Comment (elopezcastro) [reply legend-engine-executionPlan-execution-store-relational-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/relational/connection/ds/state/ConnectionStateManager.java:229]: done

File: legend-engine-executionPlan-execution-store-relational-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/relational/connection/ds/state/ConnectionStateManager.java [hunk 13]
```diff
@@ -141,101 +167,148 @@ private static ConnectionStateManager getInstanceImpl(Clock clock)
...
-    public ConnectionState getState(String poolName)
+    public DataSourceWithStatistics get(String poolName)
     {
-        return this.stateByPool.get(poolName);
+        return this.connectionPools.get(poolName);
     }
 
-    // Synchronizes using concurrent map's locks
-    public void atomicallyRemove(String poolName, ConnectionState expectedStateValue)
+    private void atomicallyRemovePool(String pool, DataSourceStatistics expectedState)
```
[the remark this answers is not in the document]
Comment (elopezcastro) [reply legend-engine-executionPlan-execution-store-relational-connection/src/main/java/org/finos/legend/engine/plan/execution/stores/relational/connection/ds/state/ConnectionStateManager.java:206]: dpne

**Review by epsstan — APPROVED** [review]

**Pull request merged**